### PR TITLE
[JAMES-3544] UploadRepository contract & InMemory implementationtemp

### PR DESCRIPTION
**Pull request by vttranlina**
Ref: https://issues.apache.org/jira/browse/JAMES-3544

https://github.com/linagora/james-project/issues/4349

### CODE REVIEW

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/exception/UploadNotFoundException.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.api.exception;
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/exception/UploadNotFoundException.java:20]: Why do we mix java and scala in the same package?

Can't we just write scala here?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import eu.timepit.refined.refineV
+import org.apache.james.jmap.api.model.Upload.Size
+import org.apache.james.mailbox.model.ContentType
+import org.slf4j.{Logger, LoggerFactory}
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+object Upload {
+  private val logger: Logger = LoggerFactory.getLogger(classOf[Upload])
+  type Size = Long Refined NonNegative
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:35]: I think we already have a scala type for Size somewhere.

Time to factorize?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    size.fold(e => {
+      logger.error(s"Encountered an invalid Upload size: $e")
+      Zero
+    },
+      refinedValue => refinedValue)
+  }
+
+  def from(metaData: UploadMetaData, content: Array[Byte]): Upload =
+    Upload(uploadId = metaData.getUploadId,
+      size = Upload.sanitizeSize(metaData.getSize),
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:49]: If we have the content at hand we should trust the content size, not the metadata handed from the client.


(Should we throw if the size specified by the client is wrong?)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      contentType = metaData.getContentType,
+      content = new ByteArrayInputStream(content))
+}
+
+case class Upload(uploadId: UploadId,
+                  size: Size,
+                  contentType: ContentType,
+                  content: InputStream)
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:57]: I've seen disasters with passing statefull parameters to classes (fun fact it was for upload).

I'd prefer a supplier of stream, it makes also the responsibility of `who closes the stream` simpler... The one calling the supplier ;-)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+object UploadHashContent {
+  def from(content: Array[Byte]): UploadHashContent =
+    UploadHashContent(Hashing.sha256.hashBytes(content).toString)
+
+}
+
+case class UploadHashContent(value: String)
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:65]: What is this for?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/UploadMetaData.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+package org.apache.james.jmap.api.model;
+
+import org.apache.james.mailbox.model.ContentType;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+
+public class UploadMetaData {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/UploadMetaData.java:28]: scala to reduce the boiler plate?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Mono;
+
+public class InMemoryUploadRepository implements UploadRepository {
+
+    private static final Map<UploadId, ImmutablePair<Username, BlobId>> uploadStore = new HashMap<>();
+    private static final Map<UploadId, UploadMetaData> uploadMetaDataStore = new HashMap<>();
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:52]: Why do we need two maps? Can't we get one?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        return Mono.zip(Mono.from(blobStore.readBytes(bucketName, blobId)),
+            Mono.just(uploadMetaDataStore.get(id)))
+            .map(bytesAndMetadata -> Upload.from(bytesAndMetadata.getT2(), bytesAndMetadata.getT1()));
+    }
+
+    private boolean userHasAccessToUpload(Username userRequest, Username owner) {
+        return userRequest.equals(owner);
+    }
+
+    private UploadContext checkUploadExits(InputStream data, Username username) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:116]: Let's not care about deduplication at this level as the blobStore does take care of it for us.

Given that UploadId is randomly generated, we will never have conflicts. 

We can thus greatly simplify the code... (And get rid of `checkUploadExits`)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/UploadId.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static UploadId from(String id) {
+        Preconditions.checkNotNull(id);
+        Preconditions.checkArgument(!id.isEmpty());
+        return new UploadId(id);
+    }
+
+    public static UploadId from(byte[] bytes) {
+        Preconditions.checkNotNull(bytes);
+        return new UploadId(Hashing.sha256().hashBytes(bytes).toString());
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/UploadId.java:46]: We should remove this: hashing and dedup is a blob store concern. We should only use the `random()` method in our generation logic.

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+            .switchIfEmpty(Mono.error(() -> new UploadNotFoundException(id)));
+    }
+
+    private Mono<Upload> retrieveUpload(UploadId id, BlobId blobId) {
+        return Mono.zip(Mono.from(blobStore.readBytes(bucketName, blobId)),
+            Mono.just(uploadMetaDataStore.get(id)))
+            .map(bytesAndMetadata -> Upload.from(bytesAndMetadata.getT2(), bytesAndMetadata.getT1()));
+    }
+
+    private boolean userHasAccessToUpload(Username userRequest, Username owner) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:112]: Is this method needed? Added value seems small to me...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                storeMapping(blobId, username, UploadMetaData.builder()
+                    .setContentType(contentType)
+                    .setUploadId(uploadId)
+                    .setSize(data.length)
+                    .build());
+                return uploadId;
+            });
+    }
+
+    private void storeMapping(BlobId blobId, Username username, UploadMetaData metaData) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:137]: Let's find a better name... (`storeMapping`)

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  @Test
+  def uploadShouldSuccess(): Unit = {
+    val uploadId: UploadId = SMono.fromPublisher(testee.upload(DATA, CONTENT_TYPE, USER)).block()
+
+    assertThat(SMono.fromPublisher(testee.retrieve(uploadId, USER)).block())
+      .isNotNull
+  }
+
+  @Test
+  def uploadShouldReturnSameIdWhenReUploadSameContent(): Unit = {
```
Comment (chibenwa) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala:58]: IMO not needed. Dedup should be a blob store concern...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  def sanitizeSize(value: Long): Size = {
+    val size: Either[String, Size] = refineV[NonNegative](value)
+    size.fold(e => {
+      logger.error(s"Encountered an invalid Upload size: $e")
+      Zero
+    },
+      refinedValue => refinedValue)
+  }
+
+  def from(metaData: UploadMetaData, content: Array[Byte]): Upload =
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:47]: Please don't enforce the use of a byte array at the heart of the data model...

(I prefer supplier of inutStream...)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala (hunk 3 above)
Comment (vttranlina) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:49]: `metaData.getSize` is not set by the client, It is set by the server when store upload (`content.length`). This purpose `def` is to convert data in the database to client response.

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java (hunk 8 above)
Comment (vttranlina) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:116]: is it in conflict with the wrote you did before? 
https://github.com/linagora/james-project/issues/4349#issuecomment-885495718

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java (hunk 10 above)
Comment (vttranlina) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:112]: I thought we can extend logic in the future, so here is a reserve. 
I will implement checking logic in-line.

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala (hunk 12 above)
Comment (vttranlina) [reply server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala:58]: What about ADR? 
(https://github.com/apache/james-project/pull/544/files#L22)
```
if reuploaded, the same blobId MAY be returned, but this SHOULD reset the expiry time.
```

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java (hunk 8 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:116]: MAY means that we are able to do it if we want, in no way that we have to do it ;-).

So no, no incompatibility at all ;-)

(And it is a quote of the JMAP spec serving another purpose there....)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java (hunk 10 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/upload/InMemoryUploadRepository.java:112]: I prefer having user.equals(owner) directly, I would find it clearer.

We can extract to a method later, when it makes sense IMO...

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala (hunk 12 above)
Comment (chibenwa) [reply server/data/data-jmap/src/test/java/org/apache/james/jmap/api/upload/UploadRepositoryContract.scala:58]: Idem

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala [hunk 14]
```diff
@@ -19,49 +19,38 @@
...
-object UploadHashContent {
-  def from(content: Array[Byte]): UploadHashContent =
-    UploadHashContent(Hashing.sha256.hashBytes(content).toString)
+case class UploadNotFoundException(uploadId: UploadId) extends RuntimeException(s"Upload not found $uploadId")
 
+object UploadMetaData {
+  def from(uploadId: UploadId, contentType: ContentType, size: Long, blobId: BlobId): UploadMetaData =
+    UploadMetaData(uploadId = uploadId,
+      contentType = contentType,
+      size = Size.sanitizeSize(size),
```
Comment (vttranlina) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/model/Upload.scala:48]: I got the error `No such instance method`, when I use `Size.sanitizeSize(size)` in Java class. 
weird

**Review by chibenwa — APPROVED** [review]
Looks good to me

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
`08:29:42,205 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project james-server-data-jmap: Compilation failure: Compilation failure`

Can we check that we have the scala compilation plugin?

**Comment by chibenwa** [comment]
```
08:32:16,512 [ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.4.6:compile (scala-compile-first) on project james-server-jmap-rfc-8621: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.4.6:compile (scala-compile-first) on project james-server-jmap-rfc-8621: wrap: org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1)
```

**Comment by chibenwa** [comment]
Can you have a look @vttranlina ?

**Comment by vttranlina** [comment]
> Can you have a look @vttranlina ?

yes, few moments

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Can you squash your fixups please? :)

**Comment by vttranlina** [comment]
rebase master & squash fixup

**Pull request merged**